### PR TITLE
Added OA-93 pistol and carbine

### PR DESCRIPTION
**Pull request by RarkGrames**
#### Summary
```SUMMARY: Content "Added OA-93 pistol and carbine"```

#### Purpose of change
Adding more compact weapons that use the 5.56mm ammo.

#### Describe the solution
- Added items
- Added items to item groups

#### Describe alternatives you've considered
Adding drop-in auto sear and lightning link mod compatibility, but I have no idea if they would work with these guns.

#### Testing
- Spawned and used the guns
- Checked out the item groups with the debug menu

#### Additional context
[Source 1](http://www.quarterbore.com/ar15m16/oa93.htm)
[Source 2](https://www.northwestfirearms.com/threads/olympic-arms-oa-93-pistol.41939/)

Rarity is based on that there is only a couple of them for sale on gun websites.

### CODE REVIEW

**Comment by I-am-Erk** [comment]
Don't forget to update Generic Guns.

File: data/json/items/gun/223.json [hunk 1]
```diff
@@ -325,6 +325,52 @@
...
     "magazines": [ [ "223", [ "stanag30", "stanag50", "survivor223mag" ] ] ]
   },
+  {
+    "id": "oa93",
+    "copy-from": "rifle_semi",
+    "looks_like": "skorpion_61",
+    "type": "GUN",
+    "name": { "str": "OA-93" },
+    "description": "An AR-15 derivative pistol manufactured by Olympic Arms in the nineties. The main difference compared to the AR-15 is that the recoil spring has been moved to the top of the gun, circumventing the necessity for a solid buttstock.",
+    "weight": "2023 g",
```
Comment (LaVeyanFiend) [inline data/json/items/gun/223.json]: Which weight are you using for this?

File: data/json/items/gun/223.json [hunk 2]
```diff
@@ -325,6 +325,52 @@
     "modes": [ [ "DEFAULT", "semi-auto", 1 ], [ "BURST", "3rd burst", 3 ] ],
     "magazines": [ [ "223", [ "stanag30", "stanag50", "survivor223mag" ] ] ]
   },
+  {
+    "id": "oa93",
+    "copy-from": "rifle_semi",
```
Comment (LaVeyanFiend) [inline data/json/items/gun/223.json]: Need to override the skill used if this is supposed to be a pistol

File: data/json/items/gun/223.json [hunk 3]
```diff
@@ -325,6 +325,52 @@
...
+    "looks_like": "skorpion_61",
+    "type": "GUN",
+    "name": { "str": "OA-93" },
+    "description": "An AR-15 derivative pistol manufactured by Olympic Arms in the nineties. The main difference compared to the AR-15 is that the recoil spring has been moved to the top of the gun, circumventing the necessity for a solid buttstock.",
+    "weight": "2023 g",
+    "volume": "750 ml",
```
Comment (LaVeyanFiend) [inline data/json/items/gun/223.json]: Trying to find the width of the upper on this for a more accurate for you ATM

File: data/json/items/gun/223.json [hunk 4]
```diff
@@ -325,6 +325,52 @@
...
+  },
+  {
+    "id": "oa93_carbine",
+    "copy-from": "rifle_semi",
+    "looks_like": "skorpion_61",
+    "type": "GUN",
+    "name": { "str": "OA-93 carbine" },
+    "description": "An AR-15 derivative carbine manufactured by Olympic Arms in the nineties. The main difference compared to the AR-15 is that the recoil spring has been moved to the top of the gun, circumventing the necessity for a solid buttstock.  This carbine version comes with built-in folding stock and longer barrel.",
+    "weight": "3402 g",
+    "volume": "1750 ml",
```
Comment (LaVeyanFiend) [inline data/json/items/gun/223.json]: Ditto for finding upper's width so I can get you a better volme

**Comment by Orinyau** [comment]
Isn't the lightning link specifically for the AR15, and more specifically AR15s with the older rectangular bolt. 

https://www.thefirearmblog.com/blog/2016/06/03/full-auto-on-the-cheap-how-a-lightning-connector-works/

(The part about the bolt is in the comments, confirmed by the author)

I know this PR doesn't add this, but it is something to think about. Also, this source states that there are only 300 lightning links registered, so it seems pretty rare.

**Comment by LaVeyanFiend** [comment]
Better volume math for the carbine:
https://www.cbcindustries.com/products/complete-ar-15-upper-assembly-16-5-56-x-45-bcg-chh-included-13-cbc-keymod-gen-2-ar-15-handguard-rail-2/
https://www.cbcindustries.com/products/ar-15-upper-assembly-16-5-56x45-1-7-midlength-12-cbc-arms-gen2-keymod-gen-2-ar-15-handguard-rail/

Using their width of 4.5in 

https://www.gunsamerica.com/990106554/Olympic-Arms-Pre-Ban-OA93-Carbine-LNIB.htm
Using this picture for scaling, since it has the stock extended:
image
image
120.416 square inches * 4.5in = 541.872in^3 = 8,880 mL
Using picture from your first source, with stock poking out removed:
image
image
113.519 in^2 * 4.5in = 510.8355in^3 = 8.371L

File: data/json/items/gun/223.json (hunk 1 above)
Comment (RarkGrames) [reply data/json/items/gun/223.json]: Second source for the pistol (Length/Weight: 17.00in. / 4.46lbs.).
The first source only states the OA-93 carbine weight.

**Comment by RarkGrames** [comment]
Is this going to be done for every gun? 7 liters is way out of the system of existing gun dimensions. Also, previously I remember you didn't take magazines for measuring.

**Comment by RarkGrames** [comment]
@Orinyau I didn't include the mods because I have no data about compatibility whatsoever.

**Comment by LaVeyanFiend** [comment]
> Is this going to be done for every gun?

Eventually, yes.

> 7 liters is way out of the system of existing gun dimensions.

Correct, because they're all too small(except for pistols, which are huge).

> Also, previously I remember you didn't take magazines for measuring.

Ewps, you're right. Will redo math.

**Comment by RarkGrames** [comment]
Are we supposed to give the volume with the stock folded? Installing a folding stock automatically decreases own the volume by 500 ml (?). Is it possible to check the barrel length sticking out of the body? I gave 67 ml for `barrel_length`, but it is not a very precise number.

**Comment by LaVeyanFiend** [comment]
> Are we supposed to give the volume with the stock folded? Installing a folding stock automatically decreases own the volume by 500 ml (?).

I asked tenmillimaster for advice on this one.

> Is it possible to check the barrel length sticking out of the body? I gave 67 ml for barrel_length, but it is not a very precise number.
From discussion on FS2000:
image

**Comment by LaVeyanFiend** [comment]
Fixed the magazine contributing to volume:
image
Subtract 15.414*4.5in = 1136.656(~1137) ml from both guns.
Carbine, retracted: 7234ml
Carbine, extended: 7743ml

**Comment by LaVeyanFiend** [comment]
Pistol math(w/o the magazine this time I promise 😛):
Using this picture from Olympic Arms' store(one in your source has shadow that would expand the volume):
http://www.olyarms.com/shop/media/catalog/product/cache/1/image/9df78eab33525d08d6e5fb8d27136e95/o/a/oa93.png
Chopped off magazine:
image
Results(length from your second source used for scaling):
image
59.441 in^2 * 4.5 in. width(since it's also an AR15-style receiver) = 267..4845 in^3 = 4383.28(~4383)ml

**Comment by RarkGrames** [comment]
uwu

**Comment by LaVeyanFiend** [comment]
I posted [a tutorial](https://github.com/CleverRaven/Cataclysm-DDA/issues/37374#issuecomment-580531920) so you can double-check my math on #37374.

File: data/json/items/gun/223.json [hunk 5]
```diff
@@ -333,7 +333,7 @@
     "name": { "str": "OA-93" },
     "description": "An AR-15 derivative pistol manufactured by Olympic Arms in the nineties. The main difference compared to the AR-15 is that the recoil spring has been moved to the top of the gun, circumventing the necessity for a solid buttstock.",
     "weight": "2023 g",
-    "volume": "750 ml",
+    "volume": "4383 ml",
```
Comment (LaVeyanFiend) [inline data/json/items/gun/223.json]: See edit, it was 4670ml.

**Comment by tenmillimaster** [comment]
The OA93 has long fallen out of popularity. If you're looking for compact 223's, why not just a normal AR pistol? They're far more applicable.

**Comment by Orinyau** [comment]
I don't think it was ever popular. Law enforcement didn't use it long, it was made illegal in '96 spawning some silly variants.

With as rare as it is, it doesn't hurt, but it doesn't fill in any holes either.

**Comment by RarkGrames** [comment]
> 
> 
> The OA93 has long fallen out of popularity. If you're looking for compact 223's, why not just a normal AR pistol? They're far more applicable.

To live out hillbilly power fantasies. Also, there are not many ar pistol variants without a buffer tube, so why not add the first one ever made. CDDA is the only game I encountered so far where your survival is directly proportional to the amount of loaded guns strapped to your body to draw, discharge, drop, repeat.

**Comment by tenmillimaster** [comment]
>To live out hillbilly power fantasies.

This is practically mod content then. You could add it make it impossibly super rare, I suppose, but that still overlooks the huge number of 10.5" and 7" barreled AR pistols out there.

**Comment by RarkGrames** [comment]
If you are so unrelenting about it, then I will close it. I don't have must-have on this thing. It just feels a little bit backwards that after adding a sea of marginally useful and obscure guns (which in some cases still have more weight than this one) you disapprove this on the basis of being too rare and having better alternatives.

**Comment by RarkGrames** [comment]
@tenmillimaster If that is fine with you, I can move these to the obscure guns item groups with a really low weight (so the work @LaVeyanFiend already did is not wasted). If you suggest some similar, but modern gun (small, 5.56, folding stock or stockless) that can be purchased nowadays, then I would implement that also.

**Comment by tenmillimaster** [comment]
To reiterate, I don't have an issue with adding an OA-93, I just think the 7.5" barreled AR's are far, far, far more common. I've gone ahead and done image analysis on one if you'd like to use it.

Based on the same data, I would use the width of receivers below (it's the width from centerline to the midline of the forward assist plunger tube, doubled) instead of 4.5" for your OA-93. 4.5" is a rather extreme size, rail-to rail of a handguard rather than average depth. If you multiply your OA-93's volumes by (2.0568/4.5) you should have your corrected volumes.


Below is the volume data for a typical 7.5" AR pistol.

image

**Comment by RarkGrames** [comment]
Ok, I can do that. Aside from a custom-built gun, is there any commercial model that uses this exact setup?

**Comment by tenmillimaster** [comment]
Pretty much all of them, though they often come with a "shoulder brace" I thought better implemented as a gunmod.

**Comment by RarkGrames** [comment]
I will make an entry with these stats then with an integrated gunmod on it.

**Comment by tenmillimaster** [comment]
Why would it be integrated and not default?

**Comment by RarkGrames** [comment]
So you can't install a folding or other type of stock on it.

**Comment by RarkGrames** [comment]
It is not integrated in a strict sense, but per my understanding these shoulder braces can only be used on the buffer tubes of these pistols, and since the tubes prevent the installation of other kind of stocks, making them integrated solves both problems, even though it is a bit of a hacky solution.

**Comment by tenmillimaster** [comment]
Yes, but no other AR-family thing has that same solution currently. I wouldn't bother considering it right now.

**Comment by RarkGrames** [comment]
Ok, then I'll add it as a default mod.

**Pull request closed without merging**